### PR TITLE
fix race in memfs

### PR DESCRIPTION
**Pull request by imjasonh**
This is like #6 and https://github.com/chainguard-dev/apko/pull/649 before it, but with finer-grained locking around the `.children` map, which had caused concurrent map access issues.

Locking `getNode()` didn't work because `getNode` can call itself recursively, leading to a deadlock. Instead this PR locks more granularly (painfully so!) around reads/writes to node's `.children`.

This also tests with `-race` in CI, to hopefully avoid introducing more of these.

Also fix lint findings, subsuming work in #7, to appease the linter.

If this works I'll port this to apko a la 649 as well.

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -21,4 +21,4 @@ jobs:
       - uses: chainguard-dev/actions/goimports@main
       - run: |
           go build ./...
-          go test ./...
+          go test ./... -race
```
Comment (luhring) [inline .github/workflows/build.yaml:24]: 😍 😍 😍 😍 😍

**Review by luhring — APPROVED** [review]
Thanks!

**Comment by deitch** [comment]
> If this works I'll port this to apko a la 649 as well

I think I would rather just start to consume this in apko. I will open a PR for it.

**Comment by deitch** [comment]
The race tests and lint cleanups I am on board with.

The constant locking and unlocking is painful (as you wrote above).

If the issue is that `getNode()` calls itself recursively, leading to deadlock, then the fix should be some `getNodeLock()` wrapper, that wraps the first `getNode()` call, which can call the others recursively.

If the issue is more fundamental, that we are not making calls that we need to, then we may need more locks as you did here. As far as I can tell, the lock calls you added are:

* lock before/unlock after reading each child when walking down `getNode()`. This one looks too granular; what happens if I read `/usr/foo/bar`, and I lock `usr` and read it, then lock `foo` and read it, then lock `bar`, but in between `foo` changed? That is what the original `getNode()` lock was in place (and why the function was created): to have an unchanged "get this node". Or maybe `getNode()` is the wrong place to put this entirely, and (since it is private anyways), `getNode()` should be assumed to be thread-unsafe, and callers should lock before?
* mkdir - lock before / unlock after adding a node to `anode.children`
* mkdirall - lock before/unlock after adding each child node when walking down the `mkdirall` tree
* openfile - lock before/unlock after getting the handle for the child (and possibly creating it)
* mknod - lock before/unlock after adding the node to `anode.children`
* readnod - lock before/unlock after reading the node from `node.children`
* symlink - lock before/unlock after adding the symlink node to `node.children`
* link - lock before/unlock after adding the symlink node to `node.children`, and lock before/unlock after updating the target count
* readlink - lock before/unlock after reading the link target

For the children reading, if this is the right thing, maybe we should have a method on node that adds to children/removes from children/reads children while wrapping with a lock? That would eliminate a lot to the overhead above, and make it easier to read (and remember what to do).

```go
func (n *node) addChild(name string, child *node) {
       n.Lock()
       defer n.Unlock()
       n.children[name] = child
}
func (n *node) getChild(name string) *node {
       n.Lock()
       defer n.Unlock()
      child, ok := n.children[name]
      if !ok {
        return nil
      }
      return nil
}
```

**Comment by deitch** [comment]
@imjasonh I fixed the issues that were blocking CI via #9. You need to rebase and fix the conflicts, and then we need to decide what the right way is to handle the various locks. I don't mean just the deadlock issues, but the right way to do locks for our purposes.

**Comment by imjasonh** [comment]
#9 doesn't fix the race:

```
go test ./... -race
?       github.com/chainguard-dev/go-apk/pkg/logger     [no test files]
?       github.com/chainguard-dev/go-apk/pkg/signature  [no test files]
==================
WARNING: DATA RACE
Read at 0x00c0000b2810 by goroutine 21:
  runtime.mapaccess1_faststr()
      /opt/homebrew/Cellar/go/1.20.3/libexec/src/runtime/map_faststr.go:13 +0x40c
  github.com/chainguard-dev/go-apk/pkg/fs.(*memFS).OpenFile()
      /Users/jason/git/go-apk/pkg/fs/memfs.go:215 +0x114
  github.com/chainguard-dev/go-apk/pkg/fs.(*memFS).WriteFile()
      /Users/jason/git/go-apk/pkg/fs/memfs.go:259 +0x58
  github.com/chainguard-dev/go-apk/pkg/apk.(*APK).InitKeyring.func1()
      /Users/jason/git/go-apk/pkg/apk/implementation.go:381 +0x610
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/jason/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x6c

Previous write at 0x00c0000b2810 by goroutine 20:
  runtime.mapaccess2_faststr()
      /opt/homebrew/Cellar/go/1.20.3/libexec/src/runtime/map_faststr.go:108 +0x42c
  github.com/chainguard-dev/go-apk/pkg/fs.(*memFS).OpenFile()
      /Users/jason/git/go-apk/pkg/fs/memfs.go:231 +0x308
  github.com/chainguard-dev/go-apk/pkg/fs.(*memFS).WriteFile()
      /Users/jason/git/go-apk/pkg/fs/memfs.go:259 +0x58
  github.com/chainguard-dev/go-apk/pkg/apk.(*APK).InitKeyring.func1()
      /Users/jason/git/go-apk/pkg/apk/implementation.go:381 +0x610
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/jason/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x6c

Goroutine 21 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/jason/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0x10c
  github.com/chainguard-dev/go-apk/pkg/apk.(*APK).InitKeyring()
      /Users/jason/git/go-apk/pkg/apk/implementation.go:333 +0x2dc
  github.com/chainguard-dev/go-apk/pkg/apk.TestInitKeyring()
      /Users/jason/git/go-apk/pkg/apk/implementation_test.go:140 +0x4c8
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.20.3/libexec/src/testing/testing.go:1576 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.20.3/libexec/src/testing/testing.go:1629 +0x40

Goroutine 20 (finished) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/jason/go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0x10c
  github.com/chainguard-dev/go-apk/pkg/apk.(*APK).InitKeyring()
      /Users/jason/git/go-apk/pkg/apk/implementation.go:333 +0x2dc
  github.com/chainguard-dev/go-apk/pkg/apk.TestInitKeyring()
      /Users/jason/git/go-apk/pkg/apk/implementation_test.go:140 +0x4c8
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.20.3/libexec/src/testing/testing.go:1576 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/Cellar/go/1.20.3/libexec/src/testing/testing.go:1629 +0x40
==================
```

I don't particularly care what the solution is, so long as we have race detection enabled in CI and tests pass.

**Pull request closed without merging**